scylla-cdc/src/lib.rs

```diff
@@ -5,9 +5,440 @@ pub mod stream_generations;
 
 #[cfg(test)]
 mod tests {
-    #[test]
-    fn it_works() {
-        let result = 2 + 2;
-        assert_eq!(result, 4);
+    use crate::consumer::*;
+    use anyhow::{anyhow, Result};
+    use async_trait::async_trait;
+    use futures::StreamExt;
+    use itertools::Itertools;
+    use scylla::frame::response::result::CqlValue;
+    use scylla::frame::value::{Value, ValueTooBig};
+    use scylla::prepared_statement::PreparedStatement;
+    use scylla::{Session, SessionBuilder};
+    use std::collections::{HashMap, VecDeque};
+    use std::hash::Hash;
+    use std::sync::Arc;
+    use tokio::sync::Mutex;
+
+    type OperationsMap = Arc<Mutex<HashMap<PrimaryKeyValue, VecDeque<Operation>>>>;
+
+    // The driver's CqlValue cannot be used as HashMap key,
+    // because it doesn't have the Eq trait.
+    #[derive(Debug, Eq, PartialEq, Hash)]
+    enum PrimaryKeyValue {
+        // Name consistency with CqlValue from the driver is recommended.
+        Int(i32),
+        Text(String),
+        List(Vec<PrimaryKeyValue>),
+        ComplexPK(Vec<PrimaryKeyValue>),
+    }
+
+    impl Value for PrimaryKeyValue {
+        fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+            self.to_cql().serialize(buf)?;
+
+            Ok(())
+        }
+    }
+
+    impl PrimaryKeyValue {
+        pub fn from_cql(cql_val: &CqlValue) -> Option<PrimaryKeyValue> {
+            match cql_val {
+                CqlValue::Int(x) => Some(PrimaryKeyValue::Int(*x)),
+                CqlValue::Text(s) => Some(PrimaryKeyValue::Text(s.clone())),
+                CqlValue::List(v) => Some(PrimaryKeyValue::List(
+                    // [FIXME]: A monad-like construction would be nice to avoid unwrapping without much code.
+                    v.iter()
+                        .map(|x| PrimaryKeyValue::from_cql(x).unwrap())
+                        .collect(),
+                )),
```

Comment:
Is there a nice way to map these lists? My intention was: if any value in the list can't be converted into `PrimaryKeyValue`, return `None`.

Reply:
You [can](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#impl-FromIterator%3COption%3CA%3E%3E) collect an `Iterator<Item = Option<T>>` to `Option<Vec<T>>`. This should work (I hope, I didn't try compiling it):

```rust
v.iter()
    .map(PrimaryKeyValue::from_cql)
    .collect::<Option<Vec<_>>()?
```

This trick also works with collecting [`Result`s](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#impl-FromIterator%3CResult%3CA%2C%20E%3E%3E).